genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+
+        // Unique id to store the resource on local disk
+        final String resourceCacheId = getResourceCacheId(sourceFileUri);
+
+        // Get a handle to the resource
+        final Resource resource = resourceLoader.getResource(uriString);
+
+        if (!resource.exists()) {
+            throw new DownloadException("Resource not found: " + uriString);
+        }
+
+        final long resourceLastModified = resource.lastModified();
+
+        //Create the resource version dir in cache if it does not exist
+        final File cacheResourceVersionDir = getCacheResourceVersionDir(
+            resourceCacheId,
+            resourceLastModified
+        );
+
+        createDirectoryStructureIfNotExists(cacheResourceVersionDir);
+
+        //Handle to the resource cached locally
+        final File cachedResourceVersionDataFile = getCacheResourceVersionDataFile(
+            resourceCacheId,
+            resourceLastModified
+        );
+
+        try (
+            final FileChannel lockFileChannel = getLockFileChannel(
+                resourceCacheId,
+                resourceLastModified
+            );
+            final Lock lock = lockingFactory.getLock(lockFileChannel);
+        ) {
+            lock.lock();
+            if (!cachedResourceVersionDataFile.exists()) {
+                log.info(
+                    "Cache miss: {} (id: {})",
+                    uriString,
+                    resourceCacheId
+                );
+
+                // Download the resource into the cache
+                final File cachedResourceVersionDownloadFile = getCacheResourceVersionDownloadFile(
+                    resourceCacheId,
+                    resourceLastModified
+                );
+                try (
+                    final InputStream in = resource.getInputStream();
+                    final OutputStream out = new FileOutputStream(cachedResourceVersionDownloadFile)
+                ) {
+                    FileCopyUtils.copy(in, out);
+                    Files.move(cachedResourceVersionDownloadFile, cachedResourceVersionDataFile);
+                }
+            } else {
+                log.info(
+                    "Cache hit: {} (id: {})",
+                    uriString,
+                    resourceCacheId
+                );
+            }
+
+            Files.copy(cachedResourceVersionDataFile, destinationFile);
+        }
+
+        //Clean up any older versions in a different thread
+        oldResourcesCleanUpService.submit(() -> {
+            try {
+                cleanUpOlderResourceVersions(resourceCacheId, resourceLastModified);
+            } catch (IOException e) {
+                log.error(
+                    "Error cleaning up old resource resourceCacheId - {}, resourceLastModified - {}",
+                    resourceCacheId,
+                    resourceLastModified,
+                    e
+                );
+            }
+        });
+    }
+
+    private FileChannel getLockFileChannel(
```

Comment:
why `getLock`?

Reply:
Bad name. Will fix.

Reply:
I spoke too soon. On looking at the code, the name makes sense. Its a helper to get a fchannel for the lock file that is created.

Reply:
This method returns a `fileChannel`. How that file channel is used is a different matter.

Reply:
Its internally gettting the file channel for a specific file - `lock` file we create. Hence the name